### ISSUE TITLE
Taxonomy term IDs not the same as what Wordpress actually wants?

### ISSUE THREAD

**Issue by jdolan** [WP-API/WP-API#1395]
I'm seeing some very strange behavior when creating and then associating terms via the REST API:

 * The API will create the new term as requested and return a well-formed response including the new term ID.
 * Attempting to associate the new term using the returned ID to an existing post will succeed as well (HTTP 201).
 * However, when viewing the post in WP-Admin, some of the term associations are either missing or entirely wrong.

What's really strange is that I can import a term via the API and then go search for it in WP-Admin as well. The ID that I got back in my REST response is NOT the term ID that Wordpress is using to render the Edit form for the term. That seems a red flag to me, but I'm not sure how or why it's happening.

Attached are screenshots of importing the term, associating the term, searching for the term against the REST API -- these are all in agreement with each other.

![screen shot 2015-07-11 at 1 50 26 pm](https://cloud.githubusercontent.com/assets/643118/8634836/5c4406b4-27d7-11e5-8569-b9565d331183.png)
![screen shot 2015-07-11 at 2 05 33 pm](https://cloud.githubusercontent.com/assets/643118/8634838/697a5b80-27d7-11e5-9775-fd90a95aa364.png)
![screen shot 2015-07-11 at 2 19 15 pm](https://cloud.githubusercontent.com/assets/643118/8634851/fc4c0ff8-27d7-11e5-88a8-08f046b292e3.png)
![screen shot 2015-07-11 at 1 53 00 pm](https://cloud.githubusercontent.com/assets/643118/8634840/784d96f4-27d7-11e5-9c90-78514d77fd95.png)

The last two screenshots are from the WP-Admin, and they clearly show that "Women's Tennis" is not associated to the Post, and the term ID is also different (27564 != 27580).

![screen shot 2015-07-11 at 1 56 15 pm](https://cloud.githubusercontent.com/assets/643118/8634841/7cb7723c-27d7-11e5-8371-cb4e8fb09216.png)
![screen shot 2015-07-11 at 1 56 52 pm](https://cloud.githubusercontent.com/assets/643118/8634842/7fa41388-27d7-11e5-8bad-7dfa14c07241.png)

Any idea what's going wrong here?

**Comment by danielbachhuber** [comment]
WP-API exposes the `term_taxonomy_id`, not the `term_id`. WordPress core still uses the latter in URLs.

**Comment by jdolan** [comment]
Okay, that's fine. But why are the associations not persisting correctly?

**Comment by danielbachhuber** [comment]
> But why are the associations not persisting correctly?

Hard to say without having access to the WP instance. It seems like there might be something else going on.

**Comment by jdolan** [comment]
So I think something is genuinely wrong here, and I'd like for this issue to be looked at a bit closer. In happy-path test cases, `term_taxonomy_id` and `term_id` are often the same. However, I'm dealing with a site with thousands of terms that have been migrated over several iterations of a site, and these identifiers are sometimes not identical. For these cases, term association via the API always fails.

In looking through the code, the API does appear to conflate these two IDs:

 * `class-wp-rest-post-terms-controller::create_item` calls `wp_set_object_terms`, providing `taxonomy_term_id` as `term_id`
 * `wp_set_object_terms`, line `~3086` checks if `term_exists`, but this method requires `term_id` and not `term_taxonomy_id`
 * `term_exists` will return false, and the call silently fails here. However, the endpoint happily resolves the term itself and informs the client that all is well.
 * Worse yet, with large enough data sets, you'll actually end up associating different terms than the ones you intended.

I've looked at this for several hours, and without actually stepping through the code in a debugger, this is my best assessment as to what is going on. Do you mind taking a brief look?

**Comment by jdolan** [comment]
I can confirm that this is indeed the bug. See pull request pull request #1430.

**Comment by joehoyle** [comment]
I actually noticed this bug too in the code a couple of days ago, however due to WordPress 4.2 doing term splitting on update, I believe that means the `term_id` and `term_taxonomy_id` are always the same on WordPress 4.2+ (even with old databases). I could be wrong on that point, so do let me know if it's possible - I think perhaps where it used to be a shared term, term-tax with a shared ID was removed, perhaps you get a non-split term with term_id and term_taxonomy_id differing. Having just said that, I just considered what happens if the auto-increment between `wp_terms` and `wp_term_taxonomy` differs, presumably that means that all new terms have different `term_id` and `term_taxonomy_id`... so I don't know. Now I've confused myself.
